### DISCUSSION TITLE
Custom Subscriptions

### DISCUSSION THREAD

**Discussion by emahuni** [directus/directus#21999] [Feature Requests]
### Describe the Improvement
Ability to use the subscription system to subscribe to custom events using extensions.

### Basic Example

I was trying to subscribe to a custom event that was going to be emitted by an extension eg:

```js
emitter.emitAction('custom', { payload: 'hello world' });
```

then in a the app, a module would sbuscribe to it as

```js
const { subscription } = await client.subscribe({
    event: 'custom',
    params: { /* ...any params or query for the custom sub */ },
  });
```

Then each time the extension emits the event, that works the same way subscriptions work. The syntax maybe different according to how you may want it to work.

I am aware that there is a websocket service that is used to emit events from the core... as discussed briefly in #21998

### Motivation

I want to be able to send custom subscription events to clients that require subscription based on custom logic on the API side.

The main reason is that there is no way of preventing certain users from accessing content/collections using the content module when you want them to have access to the same collection elsewhere in the app eg using readByQuery. So I created an endpoint to return the required data and prevent the roles I don't want completely. However, the problem comes now when I want to subscribe to data changes, there is no way of doing that.

### Detailed Design

Not sure

### Must Have:
- A Be able to subscribe data for clients
- B Be able to send data to subscribers on emitter.emitAction(...)
- C Be able to configure the initial subscription with the provided data on subscription registration then send data on whatever changes the code demands.

### Drawbacks

Not sure

### Alternatives

Create a custom subscription system though one already exists

### Adoption Strategy

Documentation alongside subscriptions documentation.

### Unresolved Questions

The api is not yet clear

**Comment by br41nslug** [comment]
As mentioned in the prev issue, that `emitter.emitAction` setup cant work because it's isolated to emitting to other extension only, please adjust the FR to use the `WebSocketService` instead.

**Reply by emahuni** [reply]
Yes, let me do that now. I have a working implementation in the form of a hook. I will update everything later today. It was more of a small tweak on the existing websocket implementation to just listen to extensions emitter events then emit upstream through the websocket service to users clients.

**Reply by emahuni** [reply]
I didn't exactly copy the code, I just reimplemented sort of what it does, howbeit in a simpler way just to get proof of concept, but from what I see it's a very minor adjustment with a bit of break in the subscribe method; the collection parameter has to be moved into options since it'll be optional for those custom subscriptions.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
